changelogs/unreleased/gh-9235-assertion_in_box_wait_limbo_acked.md

```diff
@@ -0,0 +1,6 @@
+## bugfix/core
+
+* Fixed a bug when the assertion in `box_wait_limbo_acked` would fail.
+  The assertion is that the lsn of the last entry in limbo is always
+  positive after `wal_sync` (gh-9235). In the release build, the behavior
+  does not change except that the statement is now true.
```

Comment:
Nit: `(gh-9235)` should go last, after the whole message.

Are you sure nothing would change in the release build? It seems the node would try to wait for quorum with a negative lsn and the wait would succeed immediately. As a result, the node could write a PROMOTE entry for some transactions, which aren't yet present on a majority.

Reply:
Let's remove `(gh_9235)`, because it's already in the file name.
I checked what happens in the release build.
If the `replication_synchro_quorum` is set too high on the replica, then it will never be reached after the timeout is triggered, the user will receive a `TimedOut` error. If `replication_synchro_quorum` <= number of instances in the replica set, the program will immediately stop with a `Segmentation fault` error. `box_promote` has an `if (wait_lsn < 0)` after `box_wait_limbo_acked`: https://github.com/tarantool/tarantool/blob/9a922ccc772f28e285cfa9c927083270547b0610/src/box/box.cc#L2974-L2976
Thus box_promote returns -1 without ever calling `diag_set`. `lbox_ctl_promote`. `lbox_ctl_promote` will try to call `luaT_error` and there will be a null pointer dereference.
In the changelog I simply wrote that a `Segmentation fault` would occur.

Reply:
> Thus box_promote returns -1 without ever calling diag_set. lbox_ctl_promote. lbox_ctl_promote will try to call luaT_error and there will be a null pointer dereference.
In the changelog I simply wrote that a Segmentation fault would occur

Nice. Thanks for the investigation!

> Let's remove (gh_9235), because it's already in the file name.

It should stay in the changelog text. Because the user won't see the file name, only the unified changelog.

Reply:
Fixed.